### PR TITLE
[ROS 2] Add function for same-package pluginlib tests

### PR DESCRIPTION
**Pull request by sloretz**
Adds `pluginlib_enable_plugin_testing()`
This CMake function handles creating a fake install environment for
testing pluginlib plugins in the same package that built them.
This replaces manually creating this folder structure just to run unit
tests.

This is to support testing a plugin in https://github.com/ros2/urdf/pull/13 . See this commit: https://github.com/ros2/urdf/pull/13/commits/3341988acc0fe26e494fbdd974f742fdb402032c

This function works by creating fake ament resource index entries in the build space to mimic the use of the ament index for registering packages and plugins. This was already being done manually in pluginlib (see the changes to CMakeLists.txt in this PR), but I found myself wanting to do the same thing in a downstream package.

### CODE REVIEW

**Comment by nuclearsandwich** [comment]
I haven't done a full review, nor do I have time in my schedule for one but I'm in favor of this approach and have no qualms about letting it merge if it does get a reviewer.

**Review by wjwwood — APPROVED** [review]
lgtm, approach seems good to me, documentation makes sense, and I didn't see any issues with the implementation (that FOOBAR trick is weird, but seems fine).

**Comment by sloretz** [comment]
CI (build: `--packages-up-to pluginlib` test: `--packages-select pluginlib`)
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/12059/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/7082/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/9834/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/11992/)

**Comment by sloretz** [comment]
CI LGTM and the PR has been approved except for 5ad3b38a95f21edca7f077cff2ad429d2804385d which fixes a typo in a comment. That seems trivial enough of a fix to not need a second review. Merging.

File: pluginlib/cmake/pluginlib_enable_plugin_testing.cmake
Comment (Karsten1987) [inline pluginlib/cmake/pluginlib_enable_plugin_testing.cmake:169]: This requires CMake 3.15.7 
https://cmake.org/cmake/help/v3.15/command/list.html

Before that, `POP_FRONT` is not available as a list subcommand in cmake.

Comment (Karsten1987) [reply pluginlib/cmake/pluginlib_enable_plugin_testing.cmake:169]: That also breaks the build on my ubuntu 18.04, as its cmake version there is 3.10

Comment (sloretz) [reply pluginlib/cmake/pluginlib_enable_plugin_testing.cmake:169]: Fixed by https://github.com/ros/pluginlib/pull/203

**Pull request merged**
